File: app/src/main/java/com/chaitupenjudcoder/notesapp/AddNoteActivity.java
package com.chaitupenjudcoder.notesapp;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;
import androidx.databinding.DataBindingUtil;

import com.chaitupenjudcoder.notesapp.databinding.ActivityAddNoteBinding;
import com.chaitupenjudcoder.notesapp.models.Note;

public class AddNoteActivity extends AppCompatActivity {

    public static final String ID_EXTRA = "id";
    public static final String TITLE_EXTRA = "title";
    public static final String DESCRIPTION_EXTRA = "description";
    public static final String PRIORITY_EXTRA = "priority";

    private ActivityAddNoteBinding addNoteBinding;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        addNoteBinding = DataBindingUtil.setContentView(this, R.layout.activity_add_note);

        getSupportActionBar().setHomeAsUpIndicator(R.drawable.ic_close);

        Intent intent = getIntent();
        if (intent.hasExtra(ID_EXTRA)) {
            setTitle("Edit Note");
            Note note = new Note(
                    intent.getStringExtra(TITLE_EXTRA),
                    intent.getStringExtra(DESCRIPTION_EXTRA),
                    intent.getIntExtra(PRIORITY_EXTRA, 7)
            );
            addNoteBinding.executePendingBindings();

            addNoteBinding.setNote(note);
        } else {
            setTitle("Add Note");
        }


    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.menu_item, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.save_item:
                saveItem();
                return true;
            default:
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    private void saveItem() {
        String title1 = addNoteBinding.etTitle.getText().toString();
        String desc = addNoteBinding.etDescription.getText().toString();
        int priority1 = addNoteBinding.npPriority.getValue();

        if (title1.trim().isEmpty() || desc.trim().isEmpty()) {
            Toast.makeText(this, "Please Enter Title and Description", Toast.LENGTH_SHORT).show();
            return;
        }

        Intent intent = new Intent();
        intent.putExtra(TITLE_EXTRA, title1);
        intent.putExtra(DESCRIPTION_EXTRA, desc);
        intent.putExtra(PRIORITY_EXTRA, priority1);

        int id = getIntent().getIntExtra(ID_EXTRA, -1);
        if (id != -1) {
            intent.putExtra(ID_EXTRA, id);
        }

        setResult(RESULT_OK, intent);
        finish();
    }
}

File: app/src/main/java/com/chaitupenjudcoder/notesapp/AddNoteActivity.kt
package com.chaitupenjudcoder.notesapp

import android.content.Intent
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import androidx.appcompat.app.AppCompatActivity
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.lifecycleScope
import com.chaitupenjudcoder.notesapp.databinding.ActivityAddNoteBinding
import com.chaitupenjudcoder.notesapp.models.Note
import com.chaitupenjudcoder.notesapp.utils.showToast
import com.chaitupenjudcoder.notesapp.viewmodels.NoteViewModel
import kotlinx.coroutines.flow.collectLatest
import org.koin.androidx.viewmodel.ext.android.viewModel

class AddNoteActivity: AppCompatActivity() {

    companion object {
        const val NOTE_EXTRA = "note_item"
        const val NOTE_ID_EXTRA = "note_id"
    }

    private val noteViewModel: NoteViewModel by viewModel()

    private lateinit var note: Note
    private lateinit var addNoteBinding: ActivityAddNoteBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        addNoteBinding = DataBindingUtil.setContentView(this, R.layout.activity_add_note)

        supportActionBar?.setHomeAsUpIndicator(R.drawable.ic_close)

        if (intent.hasExtra(NOTE_ID_EXTRA)) {
            title = "Edit Note"
            lifecycleScope.launchWhenStarted {
                noteViewModel.getNote(id = intent.getIntExtra(NOTE_ID_EXTRA, 1)).collectLatest {
                    note = it

                    addNoteBinding.executePendingBindings()
                    addNoteBinding.note = note
                }
            }
        } else {
            title = "Add Note"
            note = Note(id = -1, "", "", 5)
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_item, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.save_item -> {
                saveItem()
                return true
            }

            else -> Unit
        }

        return super.onOptionsItemSelected(item)
    }

    private fun saveItem() {
        val title = addNoteBinding.etTitle.text.toString()
        val desc = addNoteBinding.etDescription.text.toString()
        val priority = addNoteBinding.npPriority.value

        if (title.trim().isEmpty() || desc.trim().isEmpty()) {
            showToast("Please Enter Title and Description")
            return
        }

        val intent = Intent().apply {
            var editedNote = Note(
                title = title,
                description = desc,
                priority = priority
            )

            if (note.id != -1) {
                editedNote = editedNote.copy(id = note.id)
            } else Unit

            putExtra(NOTE_EXTRA, editedNote)
        }

        setResult(RESULT_OK, intent)
        finish()
    }
}
